### ISSUE TITLE
Gradle dependency verification caches keys for too long

### ISSUE THREAD

**Issue by liutikas** [gradle/gradle#18440]
Gradle dependency verification caches keys for too long even after keyring is deleted locally.

Note this set up uses `gradle/verification-metadata.xml` and
`gradle/verification-keyring.keys` to enable fully offline signature
verification.

1. Run `./gradlew jar`, you'll see success
2. `rm gradle/verification-keyring.keys`
3. Run `./gradlew jar`

### Expected Behavior

Step 3 fails, because we set `<key-servers enabled="false"/>` but
there is no offline keyring.

### Current Behavior
Step 3 succeeds.

Even running `./gradlew jar --rerun-tasks --refresh-keys --refresh-dependencies --no-build-cache`
passes. `./gradlew --stop` also has no effect.

The only workaround I have found so far is to specify a custom `GRADLE_USER_HOME`,
for example `GRADLE_USER_HOME=/tmp/foo ./gradlew jar` -> this will fail.

### Context
Working on enabling Gradle dependency verification for our project.

### Steps to Reproduce 
https://github.com/liutikas/gradle-bug-verification-cache
See instructions in README.md

### Your Environment
N/A

**Comment by asodja** [comment]
It looks like caching does not check if keyring exists or not in case there exists a signature for the jar. And signatures are cached next to the jar, e.g.:
```
For file:
    ~/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-scripting-compiler-embeddable/1.5.31/c6fcd3b2fc5229a2544f4324061574d96b734d9f/kotlin-scripting-compiler-embeddable-1.5.31.jar
    
Signature:
    ~/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-scripting-compiler-embeddable/1.5.31/2ed47919caaa0f84514b4ddd253fcbb126a35392/kotlin-scripting-compiler-embeddable-1.5.31.jar.asc
```
I would even say that because of cache this is "ok" behavior. Since it works in a similar way as jar caching. 

What is interesting is that even: `./gradlew clean jar --refresh-keys --refresh-dependencies --no-build-cache` does not fail. Seems like there is a lot of different cache levels involved here...

**Comment by liutikas** [comment]
> I would even say that because of cache this is "ok" behavior. Since it works in a similar way as jar caching.

Don't you think `gradle/verification-keyring.keys` should be considered an input to dependency fetching and thus should invalidate it?

> What is interesting is that even: ./gradlew clean jar --refresh-keys --refresh-dependencies --no-build-cache does not fail. Seems like there is a lot of different cache levels involved here...

I agree on this part, this is the most important bit as I cannot figure out how to "clear" the cache.

**Comment by asodja** [comment]
These .asc files in cache are anyway from maven (or any other) repository, my mistake. Yeah, probably deleting `gradle/verification-keyring.keys` should affect verification in some way.

**Comment by asodja** [comment]
Fixed with: #19703 and will be in 7.4-RC2.

Check also:
https://github.com/gradle/gradle/issues/19663#issuecomment-1023100613.
